File: Binary-Tree-Preorder-Traversal/Binary-Tree-Preorder-Traversal.py


‘’’
 Definition for a  binary tree node
 class TreeNode:
     def __init__(self, x):
         self.val = x
         self.left = None
         self.right = None
‘’’
class Solution:
    # @param root, a tree node
    # @return a list of integers
    def preorderTraversal(self, root):
        ans=[]
        if root==None:
            return ans
        ans.append(root.val);
        ans.extend(self.preorderTraversal(root.left));
        ans.extend(self.preorderTraversal(root.right));
        return ans
 

File: solutions/Binary-Tree-Preorder-Traversal/Binary-Tree-Preorder-Traversal.cpp

/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode *root) {
        stack<TreeNode*>st;
        vector<int>ans;
        while(!st.empty() || root)
        {
            if(root)
            {
                ans.push_back(root->val);
                if(root->right)
                    st.push(root->right);
                root=root->left;
            }
            else
            {
                root=st.top();
                st.pop();
            }
        }
        return ans;
    }
};


 